hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -489,13 +489,29 @@ public EntityEntry addEntry(
 
 		final EntityEntry e;
 
-		if( (entity instanceof ManagedEntity) &&  ((ManagedEntity) entity).$$_hibernate_getEntityEntry() != null && status == Status.READ_ONLY) {
-			e = ((ManagedEntity) entity).$$_hibernate_getEntityEntry();
-			e.setStatus( status  );
+		/*
+		Important: 		the following instanceof checks and castings are intentional.
+						Please do *not* refactor to make calls through the EntityEntryFactory
+						interface. Refactoring will results in polymorphic call sites which will
+						severely impact performance.
+		*/
+		if (persister.getEntityEntryFactory() instanceof MutableEntityEntryFactory){
+			e = ((MutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
+					status,
+					loadedState,
+					rowId,
+					id,
+					version,
+					lockMode,
+					existsInDatabase,
+					persister,
+					disableVersionIncrement,
+					lazyPropertiesAreUnfetched,
+					this
+			);
 		}
 		else {
-			final EntityEntryFactory entityEntryFactory = persister.getEntityEntryFactory();
-			e = entityEntryFactory.createEntityEntry(
+			e = ((ImmutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
 					status,
```

Comment:
Understood, but I think a comment about this should be added to the code to help who's going to read (and potentially modifyi) in the future.